File: firmware/src/sensors.cpp
/**
 * Kaldor IIoT - Sensor Implementation
 */

#include "sensors.h"
#include "config.h"
#include <math.h>

SensorManager::SensorManager()
    : accel(12345), dht(DHT_PIN, DHT_TYPE), readingIndex(0), numReadings(0),
      bbw_sum(0), bbw_sum_sq(0), current_min(9999), current_max(0) {

    for (int i = 0; i < 100; i++) {
        bbwReadings[i] = 0;
    }
}

bool SensorManager::begin() {
    bool success = true;

    // Initialize ultrasonic sensor pins
    pinMode(ULTRASONIC_TRIG, OUTPUT);
    pinMode(ULTRASONIC_ECHO, INPUT);
    digitalWrite(ULTRASONIC_TRIG, LOW);

    // Initialize DHT sensor
    dht.begin();
    Serial.println("  ✓ Temperature sensor initialized");

    // Initialize accelerometer
    if (!accel.begin()) {
        Serial.println("  ✗ ADXL345 not found");
        success = false;
    } else {
        accel.setRange(ADXL345_RANGE_16_G);
        Serial.println("  ✓ Accelerometer initialized");
    }

    // Perform self-test
    if (!selfTest()) {
        Serial.println("  ⚠ Sensor self-test failed");
        success = false;
    } else {
        Serial.println("  ✓ Sensor self-test passed");
    }

    return success;
}

SensorData SensorManager::read() {
    SensorData data;
    data.timestamp = millis();

    // Read BBW from ultrasonic sensor
    data.bbw = readUltrasonic();

    // Apply calibration
    data.bbw = (data.bbw + BBW_CALIBRATION_OFFSET) * BBW_CALIBRATION_SCALE;

    // Update rolling statistics
    bbwReadings[readingIndex] = data.bbw;
    readingIndex = (readingIndex + 1) % 100;
    if (numReadings < 100) numReadings++;

    updateStatistics();

    // Read other sensors at lower frequency
    static unsigned long lastSlowRead = 0;
    if (millis() - lastSlowRead > 1000) {
        lastSlowRead = millis();
        data.temperature = readTemperature();
        data.vibration = readVibration();
    }

    data.quality = calculateQuality();

    return data;
}

SensorData SensorManager::getAggregated() {
    SensorData data;
    data.timestamp = millis();

    data.bbw = current_avg;
    data.bbw_min = current_min;
    data.bbw_max = current_max;
    data.bbw_stddev = current_stddev;
    data.temperature = readTemperature();
    data.vibration = readVibration();
    data.quality = calculateQuality();

    return data;
}

float SensorManager::readUltrasonic() {
    // Send trigger pulse
    digitalWrite(ULTRASONIC_TRIG, LOW);
    delayMicroseconds(2);
    digitalWrite(ULTRASONIC_TRIG, HIGH);
    delayMicroseconds(10);
    digitalWrite(ULTRASONIC_TRIG, LOW);

    // Read echo pulse
    long duration = pulseIn(ULTRASONIC_ECHO, HIGH, 30000); // 30ms timeout

    if (duration == 0) {
        return -1; // Measurement failed
    }

    // Calculate distance in mm
    // Speed of sound = 343 m/s = 0.343 mm/μs
    // Distance = (duration / 2) * 0.343
    float distance = (duration / 2.0) * 0.343;

    return distance;
}

float SensorManager::readTemperature() {
    float temp = dht.readTemperature();

    if (isnan(temp)) {
        return -999; // Error value
    }

    return temp;
}

float SensorManager::readVibration() {
    sensors_event_t event;
    accel.getEvent(&event);

    // Calculate magnitude of acceleration
    float magnitude = sqrt(
        event.acceleration.x * event.acceleration.x +
        event.acceleration.y * event.acceleration.y +
        event.acceleration.z * event.acceleration.z
    );

    // Subtract gravity (9.8 m/s²)
    magnitude = abs(magnitude - 9.8);

    return magnitude;
}

uint8_t SensorManager::calculateQuality() {
    if (numReadings < 10) {
        return 50; // Not enough data
    }

    // Quality based on:
    // 1. Standard deviation (lower is better)
    // 2. Number of valid readings
    // 3. Sensor health

    uint8_t quality = 100;

    // Penalize high variability
    if (current_stddev > 5.0) {
        quality -= 20;
    } else if (current_stddev > 2.0) {
        quality -= 10;
    }

    // Penalize if we have invalid readings
    int invalidCount = 0;
    for (int i = 0; i < numReadings; i++) {
        if (bbwReadings[i] < 0) {
            invalidCount++;
        }
    }
    quality -= (invalidCount * 100) / numReadings;

    return max(0, min(100, quality));
}

void SensorManager::updateStatistics() {
    if (numReadings == 0) return;

    // Calculate mean
    bbw_sum = 0;
    bbw_sum_sq = 0;
    current_min = 9999;
    current_max = -9999;

    int validCount = 0;
    for (int i = 0; i < numReadings; i++) {
        float value = bbwReadings[i];
        if (value > 0) { // Valid reading
            bbw_sum += value;
            bbw_sum_sq += value * value;
            current_min = min(current_min, value);
            current_max = max(current_max, value);
            validCount++;
        }
    }

    if (validCount > 0) {
        current_avg = bbw_sum / validCount;

        // Calculate standard deviation
        float variance = (bbw_sum_sq / validCount) - (current_avg * current_avg);
        current_stddev = sqrt(max(0.0f, variance));
    } else {
        current_avg = 0;
        current_stddev = 0;
        current_min = 0;
        current_max = 0;
    }
}

void SensorManager::calibrate() {
    Serial.println("Starting calibration...");
    Serial.println("Please ensure BBW is at known reference (100mm)");
    delay(5000);

    // Take multiple readings
    float sum = 0;
    int count = 0;

    for (int i = 0; i < 100; i++) {
        float reading = readUltrasonic();
        if (reading > 0) {
            sum += reading;
            count++;
        }
        delay(100);
    }

    if (count > 0) {
        float avgReading = sum / count;
        float calibrationFactor = 100.0 / avgReading;

        Serial.printf("Calibration complete:\n");
        Serial.printf("  Average reading: %.2f mm\n", avgReading);
        Serial.printf("  Calibration factor: %.4f\n", calibrationFactor);
        Serial.printf("  Update BBW_CALIBRATION_SCALE to %.4f in config.h\n", calibrationFactor);
    } else {
        Serial.println("Calibration failed - no valid readings");
    }
}

bool SensorManager::selfTest() {
    bool success = true;

    // Test ultrasonic sensor
    float ultrasonicReading = readUltrasonic();
    if (ultrasonicReading < 0 || ultrasonicReading > 1000) {
        Serial.println("  ✗ Ultrasonic sensor test failed");
        success = false;
    }

    // Test temperature sensor
    float tempReading = readTemperature();
    if (tempReading < -50 || tempReading > 100) {
        Serial.println("  ✗ Temperature sensor test failed");
        success = false;
    }

    // Test accelerometer
    sensors_event_t event;
    if (!accel.getEvent(&event)) {
        Serial.println("  ✗ Accelerometer test failed");
        success = false;
    }

    return success;
}

File: firmware-zig/src/sensors.zig
// SPDX-License-Identifier: MPL-2.0
//! Kaldor IIoT - Sensor Management
//!
//! Manages multi-sensor data acquisition:
//! - Ultrasonic distance (Back Beam Width)
//! - Temperature (DHT22)
//! - Vibration (ADXL345 accelerometer)

const std = @import("std");
const config = @import("config.zig");
const hal = @import("hal.zig");

/// Sensor reading with all measurements
pub const SensorData = struct {
    /// Back Beam Width in millimeters
    bbw: f32 = 0,
    bbw_min: f32 = std.math.floatMax(f32),
    bbw_max: f32 = 0,
    bbw_stddev: f32 = 0,
    /// Signal quality (0-100)
    quality: u8 = 0,
    /// Temperature in Celsius
    temperature: f32 = 0,
    /// Vibration magnitude (m/s²)
    vibration: f32 = 0,
    /// Timestamp
    timestamp: u32 = 0,
};

/// Manages all sensor subsystems
pub const SensorManager = struct {
    // Ultrasonic state
    ultrasonic_initialized: bool = false,

    // Temperature state
    dht_initialized: bool = false,
    last_temp_read: u32 = 0,
    cached_temperature: f32 = 0,

    // Accelerometer state
    accel_initialized: bool = false,

    // Aggregation state
    sample_count: u32 = 0,
    bbw_sum: f64 = 0,
    bbw_sum_sq: f64 = 0,
    bbw_min: f32 = std.math.floatMax(f32),
    bbw_max: f32 = 0,
    last_bbw: f32 = 0,

    const Self = @This();

    pub fn init() Self {
        return Self{};
    }

    /// Initialize all sensors
    pub fn begin(self: *Self) bool {
        var all_ok = true;

        // Initialize ultrasonic sensor
        hal.gpio_set_output(config.ULTRASONIC_TRIG);
        hal.gpio_set_input(config.ULTRASONIC_ECHO);
        self.ultrasonic_initialized = true;

        // Initialize DHT sensor
        hal.gpio_set_input(config.DHT_PIN);
        self.dht_initialized = true;

        // Initialize accelerometer via I2C
        if (!self.initAccelerometer()) {
            all_ok = false;
        }

        return all_ok;
    }

    /// Initialize ADXL345 accelerometer
    fn initAccelerometer(self: *Self) bool {
        // Check WHO_AM_I register (0x00 should return 0xE5)
        var buf: [1]u8 = undefined;
        if (!hal.i2c_read(config.ADXL345_ADDR, 0x00, &buf)) {
            return false;
        }
        if (buf[0] != 0xE5) {
            return false;
        }

        // Set measurement mode (POWER_CTL register 0x2D)
        const power_ctl = [_]u8{ 0x2D, 0x08 };
        if (!hal.i2c_write(config.ADXL345_ADDR, &power_ctl)) {
            return false;
        }

        // Set data format (DATA_FORMAT register 0x31) - +/- 16g, full resolution
        const data_fmt = [_]u8{ 0x31, 0x0B };
        if (!hal.i2c_write(config.ADXL345_ADDR, &data_fmt)) {
            return false;
        }

        self.accel_initialized = true;
        return true;
    }

    /// Read all sensors
    pub fn read(self: *Self) SensorData {
        var data = SensorData{
            .timestamp = hal.millis(),
        };

        // Read ultrasonic (BBW)
        if (self.ultrasonic_initialized) {
            const distance = self.readUltrasonic();
            data.bbw = distance + config.BBW_OFFSET;
            data.quality = self.calculateQuality(distance);

            // Update aggregation
            self.sample_count += 1;
            self.bbw_sum += distance;
            self.bbw_sum_sq += distance * distance;
            if (distance < self.bbw_min) self.bbw_min = distance;
            if (distance > self.bbw_max) self.bbw_max = distance;
            self.last_bbw = distance;
        }

        // Read temperature (cached, DHT is slow)
        if (self.dht_initialized) {
            const now = hal.millis();
            if (now - self.last_temp_read >= 2000) { // DHT22 needs 2s between reads
                self.cached_temperature = self.readTemperature();
                self.last_temp_read = now;
            }
            data.temperature = self.cached_temperature + config.TEMP_OFFSET;
        }

        // Read accelerometer
        if (self.accel_initialized) {
            data.vibration = self.readVibration();
        }

        return data;
    }

    /// Get aggregated data and reset
    pub fn getAggregated(self: *Self) SensorData {
        var data = SensorData{
            .timestamp = hal.millis(),
            .temperature = self.cached_temperature + config.TEMP_OFFSET,
        };

        if (self.sample_count > 0) {
            const n: f64 = @floatFromInt(self.sample_count);
            const mean = self.bbw_sum / n;
            const variance = (self.bbw_sum_sq / n) - (mean * mean);
            const stddev = @sqrt(@max(0, variance));

            data.bbw = @floatCast(mean + config.BBW_OFFSET);
            data.bbw_min = self.bbw_min + config.BBW_OFFSET;
            data.bbw_max = self.bbw_max + config.BBW_OFFSET;
            data.bbw_stddev = @floatCast(stddev);
            data.quality = self.calculateQuality(@floatCast(mean));
        }

        if (self.accel_initialized) {
            data.vibration = self.readVibration();
        }

        // Reset aggregation
        self.sample_count = 0;
        self.bbw_sum = 0;
        self.bbw_sum_sq = 0;
        self.bbw_min = std.math.floatMax(f32);
        self.bbw_max = 0;

        return data;
    }

    /// Read HC-SR04 ultrasonic sensor
    fn readUltrasonic(self: *Self) f32 {
        _ = self;

        // Send 10µs trigger pulse
        hal.gpio_write(config.ULTRASONIC_TRIG, false);
        hal.delay_ms(1);
        hal.gpio_write(config.ULTRASONIC_TRIG, true);
        // Note: In real implementation, need µs delay
        hal.gpio_write(config.ULTRASONIC_TRIG, false);

        // Measure echo pulse duration
        // In freestanding, use hardware timer or pulse counter
        // Distance = (pulse_duration_us * speed_of_sound) / 2
        // speed_of_sound ≈ 343 m/s = 0.0343 cm/µs

        // Placeholder: Return simulated value
        const base_distance: f32 = 150.0; // mm
        const noise = @as(f32, @floatFromInt(hal.millis() % 100)) / 100.0 - 0.5;
        return base_distance + noise * 2.0;
    }

    /// Read DHT22 temperature sensor
    fn readTemperature(self: *Self) f32 {
        _ = self;

        // DHT22 uses a custom 1-wire protocol
        // In freestanding, implement bit-banging protocol

        // Placeholder: Return simulated value
        return 25.0 + @as(f32, @floatFromInt(hal.millis() % 100)) / 100.0;
    }

    /// Read ADXL345 accelerometer and compute vibration magnitude
    fn readVibration(self: *Self) f32 {
        _ = self;

        // Read 6 bytes starting from DATAX0 (0x32)
        var buf: [6]u8 = undefined;
        if (!hal.i2c_read(config.ADXL345_ADDR, 0x32, &buf)) {
            return 0;
        }

        // Convert to signed 16-bit (little endian)
        const x: i16 = @bitCast([2]u8{ buf[0], buf[1] });
        const y: i16 = @bitCast([2]u8{ buf[2], buf[3] });
        const z: i16 = @bitCast([2]u8{ buf[4], buf[5] });

        // Convert to m/s² (scale factor for +/- 16g mode)
        const scale: f32 = 0.004 * 9.81; // 4mg/LSB * 9.81 m/s²
        const ax: f32 = @as(f32, @floatFromInt(x)) * scale;
        const ay: f32 = @as(f32, @floatFromInt(y)) * scale;
        const az: f32 = @as(f32, @floatFromInt(z)) * scale;

        // Compute magnitude (removing gravity)
        const g: f32 = 9.81;
        const az_no_g = az - g;
        return @sqrt(ax * ax + ay * ay + az_no_g * az_no_g);
    }

    /// Calculate signal quality based on distance
    fn calculateQuality(self: *const Self, distance: f32) u8 {
        _ = self;

        // Quality degrades with distance and noise
        if (distance < 20 or distance > 400) {
            return 0; // Out of range
        }

        // Good quality in optimal range
        if (distance >= 50 and distance <= 300) {
            return 95;
        }

        // Degraded quality at edges
        if (distance < 50) {
            return @intFromFloat(50 + (distance / 50.0) * 45);
        } else {
            return @intFromFloat(95 - ((distance - 300) / 100.0) * 45);
        }
    }
};

// ============================================================================
// Tests
// ============================================================================

test "SensorManager initialization" {
    const sm = SensorManager.init();
    // begin() would fail without hardware, but init() should work
    try std.testing.expect(!sm.ultrasonic_initialized);
    try std.testing.expect(!sm.dht_initialized);
}

test "quality calculation" {
    const sm = SensorManager.init();
    try std.testing.expectEqual(@as(u8, 0), sm.calculateQuality(10));
    try std.testing.expectEqual(@as(u8, 95), sm.calculateQuality(150));
    try std.testing.expectEqual(@as(u8, 0), sm.calculateQuality(500));
}
